qtglean/pings.yaml

```diff
@@ -0,0 +1,25 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# This file defines the pings that are recorded by the Glean SDK.
+# Their code APIs is automatically generated, at build time using,
+# the `glean_parser` PyPI package.
+
+---
+$schema: moz://mozilla.org/schemas/glean/pings/2-0-0
+
+vpnsession:
+  description: |
+    Data for one VPN session, defined as a user turning on and eventually
+    turning off the VPN in the app. This ping will record start/end datestamp,
+    session-based metrics, etc. It will be sent after the user turns off the
```

Comment:
> This ping will record start/end datestamp, session-based metrics, etc.

This is unecessary. The metrics added to the ping will be declared in the metrics.yaml file. If that ever changes we'd have to update here and it's likely we will forget and this will be stale.

Reply:
> It will be sent after the user turns off the

I would like to understand the reasoning behind this scheduling choice. We had discussed a different design, previously along with @birdsarah. 

What if a user has an extremely long session? We would not hear from them for a long time and maybe be at risk of hitting Glean's 1MB ping payload limit. What if the user turns the VPN off by force killing the app, then this ping is just not sent? I think w need more triggers to this submit this ping. Please take a look at the `reasons` option for pings, we may want to use that to have a way of differenting pings sent at different time queues.

Reply:
The scheduling was requested by Yeon Joo, who is doing the data engineering on this. Having to calculate the session length (as well as a couple other in-session metrics, like "was there an connection issue") she asked if we'd be able to package these session metrics all up in one ping.

We do *not* expect that this architecture captures all corner cases, and certainly this would miss an extremely long session whose payload was over 1MB*. Santiago is comfortable with this, as we believe it captures the vast majority of VPN usage, and the goal of the metrics is to be able to better understand user behavior to make decisions for the product. Additionally, I believe this architecture includes enough trip wires for us to discover if we're missing significant user behavior with this current approach (for instance, if the custom ping is missing a "session started" metric or contains multiple "session started" metrics, it implies something a little off), and thus we can change the approach down the line if we find that we're missing more than the trivial amount of user behavior we expect to.

If I'm understanding the thoughts behind your comment, it seems like you may want as precise an accounting as possible, for all situations. If that's true, then perhaps we should have a meeting w/ Santiago, as I agree with your conclusions of some missed corner cases, it's just something that we found acceptable for now (unless/until we get data to show it's a larger portion than we expect).

*Notably, this ping is *separate* from the work that will be done to capture session start/end at the daemon (or network extension layer), which is https://mozilla-hub.atlassian.net/browse/VPN-4038. This will serve as a secondary check to see roughly how many start/ends we're missing with the current planned approach, including those that start/end from mobile system settings.

Reply:
> Santiago is comfortable with this, as we believe it captures the vast majority of VPN usage, and the goal of the metrics is to be able to better understand user behavior to make decisions for the product.

The thing is, we can't be sure of this. We don't have data to back the assumption.

---

 I am missing a request for comments on this. I would like to see a proposal / tech spec / document describing the scheduling of this ping and the reasoning for it. Specifically describing what were the discarded options and why they were discarded and documenting the discussions that happened around this design.

This ping, as I understand it, will be the transport for most of our user and health metrics. It is worth spending some time dwelling on its scheduling before moving ahead with implementation, to make sure all stakeholders agree on the design. That ideally would include engineering, product and data science because we will all consume the data collected by it.

See for example the discussion on https://mozilla-hub.atlassian.net/browse/VPN-3277 for an illustration of what I have in mind when I say I think this will be the transport of health metrics as well.

If any of my assumptions are not correct, then none of this is necessary. Let me know.

Note: this doesn't need to block this PR since it doesn't implement scheduling.

Reply:
>>    Santiago is comfortable with this, as we believe it captures the vast majority of VPN usage, and the goal of the metrics is to be able to better understand user behavior to make decisions for the product.
> The thing is, we can't be sure of this. We don't have data to back the assumption.
I fully agree that we don't have data to back the assumption, but we won't have any data to know if this assumption (that several of us share, including product, who is the one requesting these metrics) is correct until we have some data coming in - it seems like a catch-22. This data work is being done with the assumption that we'll learn things that will lead us to soon do additional instrumentation, and this hunch feels like something that can be tested in the initial go of metrics.

----

I went back and found the source of "is it okay if we miss some corner cases, but track if it's a substantial amount", and it was a DM w/ Santiago about a month ago. The first public mention of this would have been the [related ticket](https://mozilla-hub.atlassian.net/browse/VPN-4095), created 3 weeks ago. 

While this has been approved by data science and product, I suppose the only engineer approval was mine, and I should have gotten wider buy in. I'll put this on hold, and write up a document for us to discuss.

I'm going to hold this PR on a resolution on this - because if we want to change this architecture, it's very likely that this ping becomes unnecessary. This ping may exist so that the VPN session start and VPN session end can be in the same ping (with the exception of corner cases), per request of data. If we're going to rearchitect it so that they may not be in the same custom ping, we likely don't need a custom ping at all.